File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/job/DefaultJobListener.java
package net.nemerosa.ontrack.service.job;

import com.codahale.metrics.MetricRegistry;
import net.nemerosa.ontrack.job.JobKey;
import net.nemerosa.ontrack.job.JobListener;
import net.nemerosa.ontrack.job.JobRunProgress;
import net.nemerosa.ontrack.job.JobStatus;
import net.nemerosa.ontrack.model.structure.NameDescription;
import net.nemerosa.ontrack.model.support.ApplicationLogEntry;
import net.nemerosa.ontrack.model.support.ApplicationLogService;
import net.nemerosa.ontrack.model.support.SettingsRepository;
import org.springframework.boot.actuate.metrics.CounterService;
import org.springframework.transaction.annotation.Transactional;

import java.util.concurrent.TimeUnit;

public class DefaultJobListener implements JobListener {

    private final ApplicationLogService logService;
    private final MetricRegistry metricRegistry;
    private final CounterService counterService;
    private final SettingsRepository settingsRepository;

    public DefaultJobListener(ApplicationLogService logService, MetricRegistry metricRegistry, CounterService counterService, SettingsRepository settingsRepository) {
        this.logService = logService;
        this.metricRegistry = metricRegistry;
        this.counterService = counterService;
        this.settingsRepository = settingsRepository;
    }

    protected String getJobTypeMetric(JobKey key) {
        return "job-type." + key.getType().getKey();
    }

    @Override
    public void onJobStart(JobKey key) {
        counterService.increment("job");
        counterService.increment(getJobTypeMetric(key));
    }

    @Override
    @Transactional
    public void onJobPaused(JobKey key) {
        settingsRepository.setBoolean(JobListener.class, key.toString(), true);
    }

    @Override
    @Transactional
    public void onJobResumed(JobKey key) {
        settingsRepository.delete(JobListener.class, key.toString());
    }

    @Override
    public void onJobEnd(JobKey key, long milliseconds) {
        metricRegistry.timer("job").update(milliseconds, TimeUnit.MILLISECONDS);
        metricRegistry.timer(getJobTypeMetric(key)).update(milliseconds, TimeUnit.MILLISECONDS);
    }

    @Override
    public void onJobError(JobStatus status, Exception ex) {
        JobKey key = status.getKey();
        counterService.increment("error.job");
        counterService.increment("error." + getJobTypeMetric(status.getKey()));
        logService.log(
                ApplicationLogEntry.error(
                        ex,
                        NameDescription.nd(
                                key.getType().toString(),
                                key.getType().getName()
                        ),
                        status.getDescription()
                ).withDetail("job.key", key.getId())
                        .withDetail("job.progress", status.getProgressText())
        );
    }

    @Override
    public void onJobComplete(JobKey key) {
        counterService.decrement("job");
        counterService.decrement(getJobTypeMetric(key));
    }

    @Override
    public void onJobProgress(JobKey key, JobRunProgress progress) {
    }

    @Override
    public boolean isPausedAtStartup(JobKey key) {
        return settingsRepository.getBoolean(
                JobListener.class,
                key.toString(),
                false
        );
    }
}

File: ontrack-service/src/main/java/net/nemerosa/ontrack/service/job/DefaultJobListener.kt
package net.nemerosa.ontrack.service.job

import io.micrometer.core.instrument.MeterRegistry
import net.nemerosa.ontrack.job.JobKey
import net.nemerosa.ontrack.job.JobListener
import net.nemerosa.ontrack.job.JobRunProgress
import net.nemerosa.ontrack.job.JobStatus
import net.nemerosa.ontrack.model.support.SettingsRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.transaction.annotation.Transactional
import java.util.concurrent.TimeUnit

open class DefaultJobListener(
    private val meterRegistry: MeterRegistry,
    private val settingsRepository: SettingsRepository
) : JobListener {

    private val logger: Logger = LoggerFactory.getLogger(DefaultJobListener::class.java)

    override fun onJobStart(key: JobKey) {
    }

    @Transactional
    override fun onJobPaused(key: JobKey) {
        settingsRepository.setBoolean(JobListener::class.java, key.toString(), true)
    }

    @Transactional
    override fun onJobResumed(key: JobKey) {
        settingsRepository.delete(JobListener::class.java, key.toString())
    }

    override fun onJobEnd(key: JobKey, milliseconds: Long) {
        meterRegistry.timer("ontrack_job_duration_ms", key.metricTags).record(milliseconds, TimeUnit.MILLISECONDS)
        meterRegistry.counter("ontrack_job_run_count", key.metricTags).increment()
    }

    override fun onJobError(status: JobStatus, ex: Exception) {
        val key = status.key
        meterRegistry.counter("ontrack_job_errors", key.metricTags).increment()
        logger.error(
            "${status.description}, key=${key.id}, progress=${status.progressText}",
            ex
        )
    }

    override fun onJobComplete(key: JobKey) {
    }

    override fun onJobProgress(key: JobKey, progress: JobRunProgress) {}

    override fun isPausedAtStartup(key: JobKey): Boolean {
        return settingsRepository.getBoolean(
            JobListener::class.java,
            key.toString(),
            false
        )
    }
}
